[JENKINS-27027] Notify the SecurityListener on authentication

Wadeck opened the pull request:
Notify the SecurityListener in case of authentication success

- **Implementation detail**: due the current version of the method, the UserDetails required for the event was not accessible. In order to stay with the same API in SecurityListener, two "protected" methods were created to split the job and let the UserDetails accessible

See [JENKINS-27027](https://issues.jenkins-ci.org/browse/JENKINS-27027).

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Internal: Add more events on authentication (success / failure) in SecurityListener

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@reviewbybees

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

Wadeck committed:
[JENKINS-27026] Notify the SecurityListener in case of Token based authentication success
- due the current version of the method, the UserDetails required for the event was not accessible. In order to stay with the same API in SecurityListener, two "protected" methods were created to split the job and let the UserDetails accessible

Wadeck committed:
- add test to ensure the SecurityListener is called for REST Token but also for regular basic auth

Wadeck commented on test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java:
```diff
@@ -40,42 +54,57 @@ public void testVariousWaysToCall() throws Exception {
         final String token = t.getApiToken();
         makeRequestWithAuthAndVerify("foo:"+token, "foo");
+        //TODO verify why there are two events "authenticated" that are triggered
+        // the whole authentication process seems to be done twice
```
Occurring only when testing.
In case of regular web login (password), rest call (pass / token), only one call is done.

oleg-nenashev commented on test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java:
```diff
@@ -155,4 +188,79 @@ public HttpResponse doIndex() {
     }
+    
+    @TestExtension
+    public static class DummySecurityListener extends SecurityListener {
```
For the future: Maybe it makes sense to move this to a separate utility class in https://github.com/jenkinsci/jenkins-test-harness/ later (and then just inherit the extension class for it). The logic may be useful in other security-relate test classes in plugins

oleg-nenashev commented on test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java:
```diff
@@ -155,4 +188,79 @@ public HttpResponse doIndex() {
+        final EventQueue<String> loggedInCalls;
+
+        public DummySecurityListener(){
+            this.authenticatedCalls = new EventQueue<>();
```
can be just initialized in the definition, and then you could keep a default constructor

Wadeck replied:
done

oleg-nenashev commented on test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java:
```diff
@@ -155,4 +188,79 @@ public HttpResponse doIndex() {
+        
+            T t = eventList.remove(eventList.size() - 1);
+            assertTrue(predicate.test(t));
+            eventList.clear();
```
Self-modifying asserts is something unexpected to me. I would advice to document or rework it if it gets moved to JTH

Wadeck replied:
Method renamed to explicitly say we remove the event after the assert

oleg-nenashev reviewed (commented):
Looks good! Just minor comments so far

Wadeck committed:
- remove the comment about the split, will be put in GitHub comment instead

Wadeck commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -318,10 +318,26 @@ public synchronized void addProperty(@Nonnull UserProperty p) throws IOException
+     *      If this user is not a valid user in the backend {@link SecurityRealm}.
+     */
+    @Restricted(NoExternalUse.class)
+    public @CheckForNull UserDetails getUserDetailsForImpersonation() throws UsernameNotFoundException {
```
This method and the following (#impersonate(UserDetails)) results of a split of the #impersonate() method in order to have access to the UserDetails. At the moment this is only used by BasicHeaderApiTokenAuthenticator to be able to trigger the event of SecurityListener without changing the API interface.

Wadeck commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -318,10 +318,25 @@ public synchronized void addProperty(@Nonnull UserProperty p) throws IOException
+    /**
+     * Provide a way to split the job of {@link #impersonate()} into two calls in order to be able
+     * to retrieve the {@link UserDetails} used for the impersonation.
+     * First part of the split.
```
comment changed and add a GitHub comment about that

Wadeck committed:
- add check for anonymous call instead of just putting a comment
- remove the constructor in the dummy
- add link to PR from Daniel to simplify a call

Wadeck commented on test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java:
```diff
@@ -155,4 +188,79 @@ public HttpResponse doIndex() {
+
+        @Override
+        protected void failedToLogIn(@Nonnull String username) {
+        
```
Subsequent PR will follow with default implementation refactoring for SecurityListener, none currently exists

Wadeck committed:
- separate the before/after to save one clear and be more explicit
- put more meaning in the assertLastEventIs method by explicitly say we will remove the last event

Wadeck committed:
- add comment about why we do not fire the "failedToAuthenticated" in the case of an invalid token (tips: it's because it could be a valid password)

Wadeck committed:
- also add the authenticated trigger on legacy filter as pointed by Ivan

Wadeck commented:
Add event trigger also on BasicAuthenticationFilter, as pointed by Ivan.

Wadeck committed:
- add support of event on CLI remoting authentication
- adjust tests by moving the helper class used to spy on events

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

kuisathaverat commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -331,10 +347,24 @@ public synchronized void addProperty(@Nonnull UserProperty p) throws IOException
+     * @see #getUserDetailsForImpersonation()
+     */
+    @Restricted(NoExternalUse.class)
+    public @Nonnull Authentication impersonate(UserDetails userDetails) {
```
Could be the user details null?

kuisathaverat replied:
anonymous only, I suppose

kuisathaverat replied:
Taking a look at SecurityListener seems like this never should happen if the user is authenticated it should have details

Wadeck replied:
Actually the change in that code is simply a split of the existing method in order to retrieve the intermediate value.
The userDetails could be null in the case LDAP plugin is misconfigured or LDAP server is not reachable, the UserMayOrMayNotException is thrown and so in this case the returned userDetails  is null.

kuisathaverat commented on core/src/main/java/jenkins/security/BasicHeaderApiTokenAuthenticator.java:
```diff
@@ -21,14 +22,23 @@
+                UserDetails userDetails = u.getUserDetailsForImpersonation();
+                Authentication auth = u.impersonate(userDetails);
+
+                SecurityListener.fireAuthenticated(userDetails);
```
I think that at this point userDetails cannot be null

https://github.com/jenkinsci/jenkins/blob/08def67a18eee51de9f3f99bc2a792fee1c160e0/core/src/main/java/jenkins/security/SecurityListener.java#L83-L96

jglick replied:
Makes [this point of Javadoc](https://github.com/jenkinsci/jenkins/blob/bcbd5ba56ad7bbed00977b2d733f60019c6af04c/core/src/main/java/jenkins/security/SecurityListener.java#L49) inaccurate.

kuisathaverat commented on core/src/main/java/hudson/security/BasicAuthenticationFilter.java:
```diff
@@ -138,7 +141,12 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
+                UserDetails userDetails = u.getUserDetailsForImpersonation();
+                Authentication auth = u.impersonate(userDetails);
+
+                SecurityListener.fireAuthenticated(userDetails);
```
I think that at this point userDetails cannot be null

https://github.com/jenkinsci/jenkins/blob/08def67a18eee51de9f3f99bc2a792fee1c160e0/core/src/main/java/jenkins/security/SecurityListener.java#L83-L96

oleg-nenashev replied:
I agree with @kuisathaverat . The null value must be handled somehow here

Wadeck replied:
Corrected by using an inner class to manage the UserDetails in such case.

Wadeck commented:
waiting on [#3077](https://github.com/jenkinsci/jenkins/pull/3077) to use the default implementation

daniel-beck commented:
I just tried with the PR build: Going through the setup wizard leaves Jenkins without a defined root URL. 

While I think this admin monitor is a good idea, Jenkins should not be showing warnings (other than "newer Jenkins is available") to someone who just went through the setup wizard.

For that reason, I strongly disagree with merging this change as is. The setup wizard should be extended to allow defining a Jenkins URL (or could perhaps infer one from the URL used to access Jenkins in the setup wizard, similar to how the form field determines its default value?).

Only then would this change make sense.

oleg-nenashev reviewed (approved):
🐝

oleg-nenashev commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -331,10 +347,24 @@ public synchronized void addProperty(@Nonnull UserProperty p) throws IOException
+     *
+     * @param userDetails Provided by {@link #getUserDetailsForImpersonation()}
+     * @see #getUserDetailsForImpersonation()
+     */
```
It would be great to document the minimal access thing directly in Javadoc

Wadeck replied:
Second sentence on the userDetails param added as requested

oleg-nenashev reviewed (dismissed):
I agree with @kuisathaverat that we should prevent NPEs

Wadeck committed:
- as mentioned Yvan, the code had some problems with null checking, so the approach is changed in order to encapsulate all that internal mechanism

Wadeck commented:
I simplified a bit the code from external POV to encapsulate the null-checking and inner userDetail class instanciation.

Please @kuisathaverat check the modifications are correct accordingly to your initial comment 👍

kuisathaverat commented:
LGTM 🐝

oleg-nenashev commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -642,4 +653,10 @@ public static CLICommand getCurrent() {
         }
     }
+
+    public static class CLIUserDetails extends User {
```
Needs Javadoc.

Wadeck replied:
JavaDoc added but also restricted on the class since it's only used by CLI and is not expected to be used by someone else.

Wadeck committed:
- add javadoc
- open the getUserDetailsForImpersonation from the User (will let the SSHD module to retrieve UserDetails from that)

Wadeck committed:
- remove single quote in log messages

jglick commented:
> LGTM

@kuisathaverat please use the approve/reject/comment gesture in GitHub, which actually tracks completed reviews in a central spot in the UI.

jglick commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1571,7 +1572,16 @@ public static String toCCStatus(Item i) {
+
+    /**
+     * Checks if the given authentication is anonymous.
+     * @since TODO
```
Probably belongs elsewhere, such as in `ACL` perhaps.

`Functions` is _not_ a general-purpose utility class. It is bound to `h` in Jelly and I think also Groovy views, and is specifically intended as a holder for bits of Java logic that Stapler views would frequently need to invoke (and that they should not inline).

jglick replied:
Comment still stands. Minimally make this method restricted.

Wadeck replied:
Exactly what Oleg adviced me to do for the moment. We did not find a better place at the moment. The `ACL` class just had a notion of anonymous authentication like `Jenkins` but no method associated.

jglick commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1571,7 +1572,16 @@ public static String toCCStatus(Item i) {
+    /**
+     * Checks if the given authentication is anonymous.
+     * @since TODO
+     */
```
Probably also link to `Jenkins#ANONYMOUS`.

jglick commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -642,4 +653,15 @@ public static CLICommand getCurrent() {
+     * The user is never anonymous since it must be authenticated to be stored in the cache
+     */
+    @Restricted(NoExternalUse.class)
+    public static class CLIUserDetails extends User {
```
Why is this even `public`? It is used only from inside the class.

Anyway, it is created only from deprecated (Remoting-based) code, so it should also be `@Deprecated`, as a reminder to us to just delete all this stuff at some point.

jglick commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -318,23 +318,70 @@ public synchronized void addProperty(@Nonnull UserProperty p) throws IOException
             // if the user no longer exists in the backend, we need to refuse impersonating this user
-            if (!ALLOW_NON_EXISTENT_USER_TO_LOGIN)
+            if(ALLOW_NON_EXISTENT_USER_TO_LOGIN){
+                LOGGER.log(Level.FINE, "The user {0} was not found in the SecurityRealm but we are required to let it passes, due to ALLOW_NON_EXISTENT_USER_TO_LOGIN", new Object[]{ id });
```
`let it pass`

jglick commented on core/src/main/java/hudson/model/User.java:
```diff
@@ -318,23 +318,70 @@ public synchronized void addProperty(@Nonnull UserProperty p) throws IOException
+            }
         } catch (DataAccessException e) {
             // seems like it's in the same boat as UserMayOrMayNotExistException
+            LOGGER.log(Level.FINE, "The user {0} retrieval just thrown a DataAccess exception with msg = {1}, so we provide minimum access", new Object[]{ id, e.getMessage() });
```
`just threw`

jglick commented on core/src/main/java/jenkins/security/BasicHeaderApiTokenAuthenticator.java:
```diff
@@ -38,6 +48,8 @@ public Authentication authenticate(HttpServletRequest req, HttpServletResponse r
             }
         }
+        
+        
```
?

jglick commented on core/src/main/java/hudson/cli/LoginCommand.java:
```diff
@@ -45,6 +46,8 @@ protected int run() throws Exception {
         ClientAuthenticationCache store = new ClientAuthenticationCache(checkChannel());
         store.set(a);
 
+        SecurityListener.fireLoggedIn(a.getName());
```
Makes [this Javadoc](https://github.com/jenkinsci/jenkins/blob/bcbd5ba56ad7bbed00977b2d733f60019c6af04c/core/src/main/java/jenkins/security/SecurityListener.java#L63) inaccurate.

jglick commented on core/src/main/java/hudson/cli/LogoutCommand.java:
```diff
@@ -27,7 +30,13 @@ protected void printUsageSummary(PrintStream stderr) {
+
         store.remove();
+
+        SecurityListener.fireLoggedOut(auth.getName());
```
Makes [this Javadoc](https://github.com/jenkinsci/jenkins/blob/bcbd5ba56ad7bbed00977b2d733f60019c6af04c/core/src/main/java/jenkins/security/SecurityListener.java#L70) inaccurate.

jglick commented on core/src/main/java/hudson/cli/CLICommand.java:
```diff
@@ -344,8 +347,16 @@ public Channel checkChannel() throws AbortException {
+
+                if(!Functions.isAnonymous(authLoadedFromCache)){
+                    UserDetails userDetails = new CLIUserDetails(authLoadedFromCache);
+                    SecurityListener.fireAuthenticated(userDetails);
```
Makes [this Javadoc](https://github.com/jenkinsci/jenkins/blob/bcbd5ba56ad7bbed00977b2d733f60019c6af04c/core/src/main/java/jenkins/security/SecurityListener.java#L50) inaccurate.

jglick commented on core/src/main/java/hudson/security/BasicAuthenticationFilter.java:
```diff
@@ -138,7 +141,12 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
+                UserDetails userDetails = u.getUserDetailsForImpersonation();
+                Authentication auth = u.impersonate(userDetails);
+
+                SecurityListener.fireAuthenticated(userDetails);
```
Makes [this Javadoc](https://github.com/jenkinsci/jenkins/blob/bcbd5ba56ad7bbed00977b2d733f60019c6af04c/core/src/main/java/jenkins/security/SecurityListener.java#L49-L50) inaccurate on two points.

jglick commented on test/src/test/java/hudson/security/CliAuthenticationTest.java:
```diff
@@ -27,6 +37,19 @@
+
+    @After
+    public void clearListener(){
+        spySecurityListener.clearPreviousCalls();
```
Unnecessary. With `@Rule JenkinsRule`, each test case gets its own set of extensions created anew, and your list of events is correctly non`static`.

jglick commented on test/src/test/java/jenkins/security/BasicHeaderProcessorTest.java:
```diff
@@ -30,6 +32,19 @@
+
+    @After
+    public void clearListener(){
+        spySecurityListener.clearPreviousCalls();
```
ditto

jglick reviewed (changes requested):
My original advice was to leave existing `SecurityListener` events untouched, and solely introduce new event types for newly fired events: CLI login/logout/authentication, API token authentication, whatever. (Which would also have obviated the need for `UserDetails` tricks IIUC, since new method signatures could include just those bits of information that we actually reliably have at various points in the code flow.) If we want to retrofit existing event methods to be fired for new reasons, that may be OK, but at least the Javadoc must be adjusted to reflect the new behaviors.

BTW I notice that `BasicAuthenticationFilter` / `LegacySecurityRealm` is not tested here, and was not tested much in any way before either. (`JenkinsRule.configureUserRealm` exists to allow tests of it but the few tests I could find that use it could just as easily be using `createDummySecurityRealm`.) I do not know how many people actually use this mode any more; perhaps we should move it to a plugin and deprecate it if we do not expect to support it. Probably presupposes a refactoring of `SecurityFilters.groovy` into plain old Java code, which I would like to do anyway.

Wadeck committed:
- basic corrections requested by Jesse

Wadeck committed:
- just another typo

Wadeck commented:
@jglick 
Initially I had no need for the different "tricky" UserDetails because I was not aware of such corner cases.  But after the special cases treated here and in the [sshd#22](https://github.com/jenkinsci/sshd-module/pull/22), I feel like trying to make just some artificial job. I am not totally satisfied with such approach.

OTOH, your approach with a specific event for each case will be problematic when a plugin wants to add its own event. Compared to the current version where the events are "generic", we lost something quite important no ?

jglick commented:
> when a plugin wants to add its own event

I am not sure what kinds of security-related events would be _defined_ by a plugin, but if there are some, it can simply define its own listener interface.

jglick reviewed (changes requested):
As before, would need to at least adjust `SecurityListener` Javadoc to reflect new behavior. I do not have a strong opinion about whether it is better to do that, or introduce new events.

Wadeck committed:
Merge branch 'master' of https://github.com/jenkinsci/jenkins into JENKINS-27027_LOG_REST_AUTH_BY_API

Wadeck committed:
- adjust the javadoc for SecurityListener events

Wadeck committed:
- add the link to Jenkins#Anonymous

Wadeck committed:
- add link (not using see)

jglick reviewed (changes requested):
Just a prior comment about the new `isAnonymous` overload.

Wadeck committed:
- update comment on the isAnonymous as we (me + Oleg) do not find a best place at the moment

Wadeck commented:
I had a discussion with Oleg about the overloaded method and we do not find a better place at the moment. I just put the `@Restricted` annotation like @jglick and @oleg-nenashev adviced.

jglick commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1576,9 +1576,10 @@ public static boolean isAnonymous() {
-     * Checks if the given authentication is anonymous, accordingly to {@link Jenkins#ANONYMOUS}
-     * @since TODO
+     * Checks if the given authentication is anonymous, accordingly to {@link Jenkins#ANONYMOUS}.
+     * TODO not the best place to put that method, but no other place at the moment
```
I suggested `ACL`, perhaps.

Wadeck replied:
Yeah ACL was a good candidate but its responsability is more about authorization instead of authentication no ?

jglick replied:
Well, it has `SYSTEM` and `impersonate`/`as` as `static` utilities, and those are about authentication primarily. In fact it also has `ANONYMOUS` (not to be confused with `Jenkins.ANONYMOUS`, which is a different type) and some others.

Wadeck replied:
You convinced me 👍. 
So I also remove the `@Restricted` and use it in `Functions`.

Wadeck committed:
- put the new method isAnonymous in ACL instead of Functions

Wadeck committed:
- little typo
- add requirement about the SecurityContext authentication

oleg-nenashev reviewed (dismissed):
IIUC it will need to be refactored with new JTH.

I am a bit aware that `loggedIn()` changes the behavior (not only for UI now), but I do not see any implementations which may be affected in OSS. Will follow-up on private-source impl in private. Please keep on-hold  for now

oleg-nenashev commented:
We have reviewed the potentially affected implementations, no real issues with the proposed behavior change. I have removed the on-hold label

Wadeck commented:
@reviewbybees done

oleg-nenashev commented:
It was waiting in the queue for long enough. 🚢 🇮🇹
